src/server/service/leaderboard/leaderboard.ts

```diff
@@ -0,0 +1,115 @@
+import { eq } from "drizzle-orm";
+import lodash from "lodash";
+
+import { db, type BBPgTransaction } from "~/server/db/index.js";
+import { updateRanks } from "~/server/db/rank.js";
+import { ranks, users } from "~/server/db/schema.js";
+import { uuidToBigInt } from "~/utils/uuid.js";
+
+import { leaderboardWorker } from "./worker-client.js";
+
+const USERS_ON_CHAIN = 112;
+const USERS_PER_SLICE = 16;
+
+interface ScoreSlice {
+  idsSlice: string[];
+  scoresSlice: string[];
+  sliceIndex: string;
+}
+
+interface TxResult {
+  isSuccess: boolean;
+  txId?: string;
+}
+
+const calculate = (tx?: BBPgTransaction) => updateRanks(tx);
+
+const sliceScores = (userIds: string[], scores: string[]): ScoreSlice[] => {
+  const slicesDividend =
+    USERS_ON_CHAIN > userIds.length ? userIds.length : USERS_ON_CHAIN;
+  const slicesAmount = Math.ceil(slicesDividend / USERS_PER_SLICE);
+
+  return lodash.range(0, slicesAmount).map((i) => {
+    const start = i * USERS_PER_SLICE;
+    const end = (i + 1) * USERS_PER_SLICE;
+
+    if (end > userIds.length) {
+      const diff = end - userIds.length;
+
+      const fieldFill = new Array(diff).fill("0field") as string[];
+      const u64Fill = new Array(diff).fill("0u64") as string[];
+
+      return {
+        idsSlice: [...userIds.slice(start, userIds.length), ...fieldFill],
+        scoresSlice: [...scores.slice(start, scores.length), ...u64Fill],
+        sliceIndex: `${i}u8`,
+      };
+    } else {
+      return {
+        idsSlice: userIds.slice(start, end),
+        scoresSlice: scores.slice(start, end),
+        sliceIndex: `${i}u8`,
+      };
+    }
+  });
+};
+
+const executeUpdateScores = async ({
+  idsSlice,
+  scoresSlice,
+  sliceIndex,
+}: ScoreSlice): Promise<TxResult> => {
+  try {
+    const txId = (await leaderboardWorker.updateScores({
+      slice: sliceIndex,
+      userIds: idsSlice,
+      scores: scoresSlice,
+    })) as string | Error;
+
+    if (txId instanceof Error) throw txId;
+
+    return { txId, isSuccess: true };
+  } catch (error) {
+    return { isSuccess: false };
```

Comment:
Should we return an empty string for the `txId` for the sake of type consistency?

Reply:
It is type consistent. The type allows `txId` to be undefined and the compiler will infer that it is undefined if the flag is false. Returning an empty string would actually make it less type consistent since you would return string in both cases